pkg/virtctl/credentials/add-key/add-key_test.go

```diff
@@ -0,0 +1,562 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/kubernetes/fake"
+	k8stesting "k8s.io/client-go/testing"
+	"k8s.io/utils/pointer"
+	"kubevirt.io/api/core"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/api"
+	"kubevirt.io/client-go/testutils"
+
+	"github.com/golang/mock/gomock"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/tests/clientcmd"
+)
+
+var _ = Describe("Credentials add-ssh-key", func() {
+	const (
+		vmName     = "test-vm"
+		secretName = "test-secret"
+		userName   = "test-user"
+	)
+
+	var (
+		kubeClient *fake.Clientset
+
+		vmi *v1.VirtualMachineInstance
+		vm  *v1.VirtualMachine
+	)
+
+	BeforeEach(func() {
+		kubeClient = fake.NewSimpleClientset()
+
+		kubeClient.Fake.PrependReactor("create", "secrets", func(action k8stesting.Action) (bool, runtime.Object, error) {
+			created, ok := action.(k8stesting.CreateAction)
+			Expect(ok).To(BeTrue())
+
+			// GenerateName is not handled by default
+			secret := created.GetObject().(*corev1.Secret)
+			if secret.GenerateName != "" {
+				secret.Name = secret.GenerateName + rand.String(6)
+			}
+			return false, secret, nil
+		})
+
+		vmi = api.NewMinimalVMI(vmName)
+		vmi.Spec.AccessCredentials = []v1.AccessCredential{{
+			SSHPublicKey: &v1.SSHPublicKeyAccessCredential{
+				Source: v1.SSHPublicKeyAccessCredentialSource{
+					Secret: &v1.AccessCredentialSecretSource{
+						SecretName: secretName,
+					},
+				},
+				PropagationMethod: v1.SSHPublicKeyAccessCredentialPropagationMethod{
+					QemuGuestAgent: &v1.QemuGuestAgentSSHPublicKeyAccessCredentialPropagation{
+						Users: []string{userName},
+					},
+				},
+			},
+		}}
+
+		vm = kubecli.NewMinimalVM(vmName)
+		vm.Namespace = metav1.NamespaceDefault
+		vm.Spec.Template = &v1.VirtualMachineInstanceTemplateSpec{
+			Spec: vmi.Spec,
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      secretName,
+				Namespace: metav1.NamespaceDefault,
+				OwnerReferences: []metav1.OwnerReference{{
+					APIVersion: v1.VirtualMachineGroupVersionKind.GroupVersion().String(),
+					Kind:       v1.VirtualMachineGroupVersionKind.Kind,
+					Name:       vm.Name,
+					UID:        vm.UID,
+					Controller: pointer.Bool(true),
+				}},
+			},
+		}
+
+		_, err := kubeClient.CoreV1().Secrets(metav1.NamespaceDefault).Create(context.Background(), secret, metav1.CreateOptions{})
+		Expect(err).ToNot(HaveOccurred())
+
+		ctrl := gomock.NewController(GinkgoT())
+		kubecli.GetKubevirtClientFromClientConfig = kubecli.GetMockKubevirtClientFromClientConfig
+		kubecli.MockKubevirtClientInstance = kubecli.NewMockKubevirtClient(ctrl)
+
+		vmiInterface := kubecli.NewMockVirtualMachineInstanceInterface(ctrl)
+		vmInterface := kubecli.NewMockVirtualMachineInterface(ctrl)
+
+		kubecli.MockKubevirtClientInstance.EXPECT().VirtualMachineInstance(metav1.NamespaceDefault).Return(vmiInterface).AnyTimes()
```

Comment:
Can we avoid the AnyTimes and us the exact number of times?

Reply:
Can you explain why? It's just a mock to return another mock. It doesn't matter how many times it's called, and maybe not all `It()` tests will call it.

Reply:
I understand